### PR TITLE
Modifying code handle swift backend

### PR DESCRIPTION
**Pull request by nguyenhuuluan434**
#### What is the purpose of this change? ####

  

The PR is create with some purpose:
  - Add '_' char to white list when encode url.
  - Unify the util check and get manifest for DLO or SLO to common function.
  - Using delete bulk during delete segments.
  - Using the atexit package to handle unexpected error during sync cache to remote backend
#### Was the change discussed in an issue or in the forum before?

[link request https://github.com/rclone/rclone/issues/4889](https://github.com/rclone/rclone/issues/4889)
#### Checklist

- [ ] I have read the [contribution guidelines](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#submitting-a-pull-request).
- [ ] I have added tests for all changes in this PR if appropriate.
- [ ] I have added documentation for the changes if appropriate.
- [ ] All commit messages are in [house style](https://github.com/rclone/rclone/blob/master/CONTRIBUTING.md#commit-messages).
- [ ] I'm done, this Pull Request is ready for review :-)

### CODE REVIEW

**Comment by ivandeex** [comment]
This change must include a unit test that simulates acting on a single segment from two rclone copies and checks that segment is neither left dangling nor prematurely deleted afterwards.

**Comment by nguyenhuuluan434** [comment]
Hi @ivandeex ,
I had added unit test for large object but this is happy case upload successful. Could you recommend any suggestion what i can cancel during run unit test to trigger the atexit.OnError to delete the uploaded segments ? 

Thanks,
Luan

**Comment by ncw** [comment]
> I had added unit test for large object but this is happy case upload successful.

Well done - looks nice :-)

> Could you recommend any suggestion what i can cancel during run unit test to trigger the atexit.OnError to delete the uploaded segments ?

What we usually do here is create a reader which gives an error after reading so many bytes. You can create one by combining the bytes reader and an error reader something like this

```go
	contents := random.String(contentSize)
￼	buf := bytes.NewBufferString(contents[:errposition])
        er := &readers.ErrorReader{Err: errors.New("potato")} // from lib/readers
        in := ioutil.NopCloser(io.MultiReader(buf, err))
￼```

**Comment by nguyenhuuluan434** [comment]
Hi @ncw ,
I had added the unit test for the case what need clean up segments.
Please help me review.
Thanks,
Luan

**Comment by nguyenhuuluan434** [comment]
Hi @ncw ,
Just remind. :-)

File: backend/swift/swift.go [hunk 1]
```diff
@@ -1127,48 +1135,77 @@ func min(x, y int64) int64 {
...
 		}
-		return shouldRetry(err)
-	})
-	if err == nil {
-		fs.Debugf(o, "Removed empty container %q", segmentsContainer)
 	}
 	return nil
 }
 
+//func removeObjects(f *Fs, container string, prefix string) (err error) {
```
Comment (ncw) [inline backend/swift/swift.go:1171]: Can you remove the dead code please.

File: backend/swift/swift.go [hunk 2]
```diff
@@ -1194,7 +1231,7 @@ func urlEncode(str string) string {
 	var buf bytes.Buffer
 	for i := 0; i < len(str); i++ {
 		c := str[i]
-		if (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '/' || c == '.' {
+		if (c == '_' || c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '/' || c == '.' {
```
Comment (ncw) [inline backend/swift/swift.go:1234]: What is this change about?

It looks wrong though - I think it should be:

```go
if (c >= '0' && c <= '9') || (c >= 'a' && c <= 'z') || (c >= 'A' && c <= 'Z') || c == '/' || c == '.'  || c == '_' {
```

File: backend/swift/swift.go [hunk 3]
```diff
@@ -276,6 +283,7 @@ var retryErrorCodes = []int{
 	401, // Unauthorized (e.g. "Token has expired")
 	408, // Request Timeout
 	409, // Conflict - various states that could be resolved on a retry
+	413, // Upload exceeds quota
```
Comment (ncw) [inline backend/swift/swift.go:286]: Is this something we want to retry? If we retry it then surely the upload will still exceed the quota?

**Review by ncw — COMMENTED** [review]
See inline for a few notes.

Note that when I run the integration tests with `go test -v` (you'll need docker installed) these don't pass, in particular the `TestIntegration/FsMkdir/FsPutChunked` tests fail indicating there is a problem with the new code :-(

I didn't figure out what - can you see the problem?

File: backend/swift/swift.go (hunk 2 above)
Comment (nguyenhuuluan434) [reply backend/swift/swift.go:1234]: my mistake :(

**Comment by nguyenhuuluan434** [comment]
Hi @ncw ,
The integration tests fail because the Swift accept dash char ("-") but in urlEncode function will encode this to %2D. I has added dash to white list char during encode.
Please help me review again.
Thanks,
Luan.

**Comment by nguyenhuuluan434** [comment]
Hi @ncw ,
Help me review, please.
Thanks,
Luân Nguyễn Hữu.

**Comment by ivandeex** [comment]
@nguyenhuuluan434 
Please mark resolved converssations

**Review by ncw — APPROVED** [review]
This is looking great now, thank you @nguyenhuuluan434 

I'm going to squash and merge now :-)

**Pull request merged**